### ISSUE TITLE
Drop/Delete Range of Keys

### ISSUE THREAD

**Issue by 256dpi** [dgraph-io/badger#922]
**TL;DR**

It would be wonderful if badger would support some sort of optimized way to handle deletion of a range of keys. Inspiration can be taken from other projects like RocksDB which implements a special range delete tombstone: https://rocksdb.org/blog/2018/11/21/delete-range.html.

**Use Case**

I'm currently working on [quasar](https://github.com/256dpi/quasar) which builds on badger to implement a persistent queue. There is a small runnable example here: https://github.com/256dpi/quasar/blob/master/example/main.go. The whole thing works pretty well and in a basic setup with one producer and one consumer the steady throughput is about 100'000 messages/second on my computer. I think with badger not being necessarily optimized for that workload it's pretty neat! 👍 

Since the queue has a fixed size, there is a "cleaner" process that periodically deletes processed messages to free space in the queue. This process iterates over the keys and removes them in using one or two transactions: https://github.com/256dpi/quasar/blob/master/ledger.go#L409.

Unfortunately, this approach is very expensive and is responsible for up to 30% of the CPU time used by the program:

<img width="1756" alt="flame-graph" src="https://user-images.githubusercontent.com/696886/60770920-a6056280-a0e0-11e9-9967-371d305c91b0.png">

_The flame graph incorrectly attributes all `badger.(*Iterator).prefetch` calls to the initial `badger.(*Iterator).Seek` call. The normal graph does not make that connection and I assume they also come from the subsequent `badger.(*Iterator).Next()` calls._

With the support of something similar to range delete tombstones in RocksDB this would be possibly a lot faster.  Also I would favor an approach that is not externally as for example `db.DropPrefix` is now. In my use case writes and deletes can happen at the same time and thus I would not want to block the database for cleanings.

I imagine an interface like the following:

```go
func (txn *Txn) DeleteRange(from, to []byte) error {}
```

The same functionality could potentially also be used to delete based on a prefix:

```go
func (txn *Txn) DeletePrefix(prefix []byte) error {
  return txn.DeleteRange(prefixStart(prefix), prefixEnd(prefix))
}
```

I didn't to any research yet on how this could be implemented in badger. Maybe somebody from the team has already thought about this. Anyway, thanks for considering this!

**Comment by vgtom** [comment]
Which parts of the code would i need to look to get started on this??

**Comment by jarifibrahim** [comment]
This looks like a useful improvement.

**Comment by minhaj-shakeel** [comment]
**Github issues have been deprecated.** 
This issue has been moved to [discuss](https://discuss.dgraph.io/t/8723). You can follow the conversation there and also subscribe to updates by changing your notification preferences.

 <img src="https://github.com/dgraph-io/benchmarks/blob/master/images/SetNotifications.png?raw=true" alt="drawing" width="40%"/>
